File: packages/android/locanara/src/main/kotlin/com/locanara/builtin/ProofreadChain.kt
package com.locanara.builtin

import com.locanara.ProofreadCorrection
import com.locanara.ProofreadResult
import com.locanara.composable.Chain
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel

/**
 * Built-in chain for grammar and spelling correction.
 *
 * ```kotlin
 * val result = ProofreadChain().run("Ths is a tset")
 * println(result.correctedText) // "This is a test"
 * ```
 */
class ProofreadChain(
    private val model: LocanaraModel = LocanaraDefaults.model
) : Chain {
    override val name = "ProofreadChain"

    override suspend fun invoke(input: ChainInput): ChainOutput {
        val prompt = BuiltInPrompts.proofread.format(
            mapOf("text" to input.text)
        )

        println("[ProofreadChain] input: ${input.text.take(200)}")
        val response = model.generate(prompt, GenerationConfig.STRUCTURED)
        val corrected = BuiltInPrompts.stripPreamble(response.text.trim())
        println("[ProofreadChain] output: $corrected")

        val corrections = detectCorrections(input.text, corrected)
        println("[ProofreadChain] corrections: ${corrections.joinToString { "${it.original} → ${it.corrected}" }}")

        val result = ProofreadResult(
            correctedText = corrected,
            corrections = corrections,
            hasCorrections = corrections.isNotEmpty()
        )

        return ChainOutput(
            value = result,
            text = corrected,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )
    }

    /** Detect corrections by comparing original and corrected text word-by-word. */
    private fun detectCorrections(original: String, corrected: String): List<ProofreadCorrection> {
        val originalWords = original.split("\\s+".toRegex())
        val correctedWords = corrected.split("\\s+".toRegex())
        val corrections = mutableListOf<ProofreadCorrection>()

        val minLen = minOf(originalWords.size, correctedWords.size)
        var searchStart = 0
        for (i in 0 until minLen) {
            if (originalWords[i] != correctedWords[i]) {
                val origClean = originalWords[i].replace(Regex("[^a-zA-Z']"), "")
                val corrClean = correctedWords[i].replace(Regex("[^a-zA-Z']"), "")
                val type = if (origClean.equals(corrClean, ignoreCase = true)) "punctuation" else "spelling"

                val startPos = original.indexOf(originalWords[i], searchStart)
                val endPos = if (startPos >= 0) startPos + originalWords[i].length else null
                if (startPos >= 0) searchStart = startPos + originalWords[i].length

                corrections.add(
                    ProofreadCorrection(
                        original = originalWords[i],
                        corrected = correctedWords[i],
                        type = type,
                        confidence = 0.9,
                        startPos = if (startPos >= 0) startPos else null,
                        endPos = endPos
                    )
                )
            }
        }
        return corrections
    }

    /** Type-safe execution that returns [ProofreadResult] directly. */
    suspend fun run(text: String): ProofreadResult {
        val output = invoke(ChainInput(text = text))
        return output.typed<ProofreadResult>()
            ?: throw IllegalStateException("Unexpected output type from ProofreadChain")
    }
}

File: packages/apple/Sources/BuiltIn/ProofreadChain.swift
import Foundation
#if canImport(FoundationModels)
import FoundationModels
#endif

/// Built-in chain for grammar and spelling correction.
///
/// Uses `@Generable` structured output on Apple Intelligence for accurate corrections
/// with a detailed corrections list. Falls back to text-based proofreading on other models.
///
/// ```swift
/// let result = try await ProofreadChain().run("Ths is a tset")
/// ```
@available(iOS 15.0, macOS 14.0, *)
public struct ProofreadChain: Chain {
    public let name = "ProofreadChain"

    private let model: any LocanaraModel

    public init(model: (any LocanaraModel)? = nil) {
        self.model = model ?? LocanaraDefaults.model
    }

    // MARK: - @Generable structured output types

    #if canImport(FoundationModels)
    @available(iOS 26.0, macOS 26.0, *)
    @Generable
    struct CorrectionOutput {
        @Guide(description: "The original incorrect word or phrase")
        var original: String
        @Guide(description: "The corrected version")
        var corrected: String
        @Guide(description: "Error type: spelling, grammar, or punctuation")
        var type: String
    }

    @available(iOS 26.0, macOS 26.0, *)
    @Generable
    struct ProofreadOutput {
        @Guide(description: "The full corrected text with all corrections applied")
        var correctedText: String
        @Guide(description: "List of individual corrections made, empty if no corrections needed")
        var corrections: [CorrectionOutput]
    }
    #endif

    // MARK: - Invoke

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        print("[ProofreadChain] input: \(input.text.prefix(200))")

        let startTime = Date()

        // Try structured generation first (Apple Intelligence)
        #if canImport(FoundationModels)
        if #available(iOS 26.0, macOS 26.0, *) {
            let prompt = """
            Proofread the following text for grammar, spelling, and punctuation errors.
            Provide the full corrected text and list each individual correction.

            Text to proofread:
            <input>\(input.text.replacingOccurrences(of: "</input>", with: ""))</input>
            """

            if let output: ProofreadOutput = try await model.generateStructured(prompt: prompt, type: ProofreadOutput.self) {
                let correctedText = output.correctedText.trimmingCharacters(in: .whitespacesAndNewlines)
                let elapsed = Int(Date().timeIntervalSince(startTime) * 1000)
                print("[ProofreadChain] output (structured): \(correctedText)")

                var searchStart = input.text.startIndex
                let corrections = output.corrections.map { c in
                    var correction = ProofreadCorrection(
                        original: c.original,
                        corrected: c.corrected,
                        type: c.type,
                        confidence: 0.9,
                        startPos: nil,
                        endPos: nil
                    )
                    if let range = input.text.range(of: c.original, range: searchStart..<input.text.endIndex) {
                        correction.startPos = input.text.distance(from: input.text.startIndex, to: range.lowerBound)
                        correction.endPos = input.text.distance(from: input.text.startIndex, to: range.upperBound)
                        searchStart = range.upperBound
                    }
                    return correction
                }

                let result = ProofreadResult(
                    correctedText: correctedText.isEmpty ? input.text : correctedText,
                    corrections: corrections,
                    hasCorrections: !corrections.isEmpty
                )

                return ChainOutput(
                    value: result,
                    text: correctedText,
                    metadata: input.metadata,
                    processingTimeMs: elapsed
                )
            }
        }
        #endif

        // Fallback: text-based proofreading
        let prompt = try BuiltInPrompts.proofread.format([
            "text": input.text
        ])

        let response = try await model.generate(prompt: prompt, config: .structured)
        let corrected = BuiltInPrompts.stripPreamble(response.text.trimmingCharacters(in: .whitespacesAndNewlines))
        print("[ProofreadChain] output (text): \(corrected)")

        let result = ProofreadResult(
            correctedText: corrected,
            corrections: [],
            hasCorrections: corrected != input.text
        )

        return ChainOutput(
            value: result,
            text: corrected,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )
    }

    /// Type-safe execution that returns `ProofreadResult` directly.
    public func run(_ text: String) async throws -> ProofreadResult {
        let output = try await invoke(ChainInput(text: text))
        guard let result = output.typed(ProofreadResult.self) else {
            throw LocanaraError.executionFailed("Unexpected output type from ProofreadChain")
        }
        return result
    }
}
